socket.url.buildpath() path protection unit test fails

I've not been able to isolate it yet, but there seems to be a discrepancy in the uppercase/lowercase hex encoding between Lua and GopherLua.

```bash
$ go test
Running test urltest.lua
testing URL parsing
testing URL building
testing absolute resolution
testing path parsing and composition
testing path protection
%2Feu%3E/%3Cdiego%3F	%2feu%3e/%3cdiego%3f
--- FAIL: TestUrl (0.01s)
        Error Trace:    gluasocket_test.go:75
	            	gluasocket_test.go:51
	Error:      	Received unexpected error:
	            	urltest.lua:18: path composition failed.
	            	stack traceback:
	            		[G]: in function 'error'
	            		urltest.lua:18: in function 'check_protect'
	            		urltest.lua:604: in main chunk
	            		[G]: ?
FAIL
exit status 1
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,7 @@
 ### Fixed
 - [PR-14](https://github.com/nubix-io/gluasocket/pull/14): `socket.core` `skip()` results were off by one; provide `settimeout()` missing return value for compatiblility with `try()`
 - [PR-15](https://github.com/nubix-io/gluasocket/pull/15): smtp sending is fixed through proper swapping of metatables on connect
+- Integrate [luasocket#3ee8951](https://github.com/diegonehab/luasocket/commit/3ee89515a0ef4852f64b13133c22aa7d3a322cfd#diff-4d7e24364dca5902525b2638230cb132) (19 Nov 2017): fixed URL parsing in url.lua: parse fragment after parsing username and password
 
 ## [0.1.1] - 2018-01-12
 ### Added

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -3,6 +3,7 @@
 - [PR-14](https://github.com/nubix-io/gluasocket/pull/14): `socket.core` `skip()` results were off by one; provide `settimeout()` missing return value for compatiblility with `try()`
 - [PR-15](https://github.com/nubix-io/gluasocket/pull/15): smtp sending is fixed through proper swapping of metatables on connect
 - Integrate [luasocket#3ee8951](https://github.com/diegonehab/luasocket/commit/3ee89515a0ef4852f64b13133c22aa7d3a322cfd#diff-4d7e24364dca5902525b2638230cb132) (19 Nov 2017): fixed URL parsing in url.lua: parse fragment after parsing username and password
+- Integrate [luasocket#2d6a0f7](https://github.com/diegonehab/luasocket/commit/2d6a0f7bda9241f827a3edbfa738603c024a423b#diff-4d7e24364dca5902525b2638230cb132) (22 Nov 2017): fixed url parsing; postpone fragment parsing after authority parsing; added test cases to test/urltest.lua; fixed reference patterns in check_protect() to upper case hex letters
 
 ## [0.1.1] - 2018-01-12
 ### Added

--- a/gluasocket_test.go
+++ b/gluasocket_test.go
@@ -50,9 +50,9 @@ func TestSocketGetAddrInfo(t *testing.T) {
 //	doTest("test_socket_error.lua", t)
 //}
 
-//func TestUrl(t *testing.T) {
-//	doTest("urltest.lua", t)
-//}
+func TestUrl(t *testing.T) {
+	doTest("urltest.lua", t)
+}
 
 // ----------------------------------------------------------------------------
 func doTest(testScript string, t *testing.T) {

--- a/socketurl/url.go
+++ b/socketurl/url.go
@@ -155,14 +155,14 @@ function _M.parse(url, default)
         parsed.authority = n
         return ""
     end)
-    -- get query string
-    url = string.gsub(url, "%?(.*)", function(q)
-        parsed.query = q
-        return ""
-    end)
     -- get fragment
     url = string.gsub(url, "#(.*)$", function(f)
         parsed.fragment = f
+        return ""
+    end)
+    -- get query string
+    url = string.gsub(url, "%?(.*)", function(q)
+        parsed.query = q
         return ""
     end)
     -- get params

--- a/socketurl/url.go
+++ b/socketurl/url.go
@@ -147,11 +147,6 @@ function _M.parse(url, default)
     if not url or url == "" then return nil, "invalid url" end
     -- remove whitespace
     -- url = string.gsub(url, "%s", "")
-    -- get fragment
-    url = string.gsub(url, "#(.*)$", function(f)
-        parsed.fragment = f
-        return ""
-    end)
     -- get scheme
     url = string.gsub(url, "^([%w][%w%+%-%.]*)%:",
         function(s) parsed.scheme = s; return "" end)
@@ -163,6 +158,11 @@ function _M.parse(url, default)
     -- get query string
     url = string.gsub(url, "%?(.*)", function(q)
         parsed.query = q
+        return ""
+    end)
+    -- get fragment
+    url = string.gsub(url, "#(.*)$", function(f)
+        parsed.fragment = f
         return ""
     end)
     -- get params

--- a/testdata/luasocket-test/urltest.lua
+++ b/testdata/luasocket-test/urltest.lua
@@ -84,6 +84,75 @@ end
 
 print("testing URL parsing")
 check_parse_url{
+    url = "scheme://user:pass$%?#wd@host:port/path;params?query#fragment",
+    scheme = "scheme", 
+    authority = "user:pass$%?#wd@host:port", 
+    host = "host",
+    port = "port",
+    userinfo = "user:pass$%?#wd",
+    password = "pass$%?#wd",
+    user = "user",
+    path = "/path",
+    params = "params",
+    query = "query",
+    fragment = "fragment"
+ }
+ check_parse_url{
+    url = "scheme://user:pass?#wd@host:port/path;params?query#fragment",
+    scheme = "scheme", 
+    authority = "user:pass?#wd@host:port", 
+    host = "host",
+    port = "port",
+    userinfo = "user:pass?#wd",
+    password = "pass?#wd",
+    user = "user",
+    path = "/path",
+    params = "params",
+    query = "query",
+    fragment = "fragment"
+ }
+ check_parse_url{
+    url = "scheme://user:pass-wd@host:port/path;params?query#fragment",
+    scheme = "scheme", 
+    authority = "user:pass-wd@host:port", 
+    host = "host",
+    port = "port",
+    userinfo = "user:pass-wd",
+    password = "pass-wd",
+    user = "user",
+    path = "/path",
+    params = "params",
+    query = "query",
+    fragment = "fragment"
+ }
+ check_parse_url{
+    url = "scheme://user:pass#wd@host:port/path;params?query#fragment",
+    scheme = "scheme", 
+    authority = "user:pass#wd@host:port", 
+    host = "host",
+    port = "port",
+    userinfo = "user:pass#wd",
+    password = "pass#wd",
+    user = "user",
+    path = "/path",
+    params = "params",
+    query = "query",
+    fragment = "fragment"
+ }
+ check_parse_url{
+    url = "scheme://user:pass#wd@host:port/path;params?query",
+    scheme = "scheme", 
+    authority = "user:pass#wd@host:port", 
+    host = "host",
+    port = "port",
+    userinfo = "user:pass#wd",
+    password = "pass#wd",
+    user = "user",
+    path = "/path",
+    params = "params",
+    query = "query",
+ }
+ check_parse_url{
     url = "scheme://userinfo@host:port/path;params?query#fragment",
     scheme = "scheme", 
     authority = "userinfo@host:port", 
@@ -601,9 +670,9 @@ check_parse_path("eu/tu", { "eu", "tu" })
 print("testing path protection")
 check_protect({ "eu", "-_.!~*'():@&=+$,", "tu" }, "eu/-_.!~*'():@&=+$,/tu")
 check_protect({ "eu ", "~diego" }, "eu%20/~diego")
-check_protect({ "/eu>", "<diego?" }, "%2feu%3e/%3cdiego%3f")
-check_protect({ "\\eu]", "[diego`" }, "%5ceu%5d/%5bdiego%60")
-check_protect({ "{eu}", "|diego\127" }, "%7beu%7d/%7cdiego%7f")
+check_protect({ "/eu>", "<diego?" }, "%2Feu%3E/%3Cdiego%3F")
+check_protect({ "\\eu]", "[diego`" }, "%5Ceu%5D/%5Bdiego%60")
+check_protect({ "{eu}", "|diego\127" }, "%7Beu%7D/%7Cdiego%7F")
 check_protect({ "eu ", "~diego" }, "eu /~diego", 1)
 check_protect({ "/eu>", "<diego?" }, "/eu>/<diego?", 1)
 check_protect({ "\\eu]", "[diego`" }, "\\eu]/[diego`", 1)

